### DISCUSSION TITLE
Plans to add type annotations?

### DISCUSSION THREAD

**Discussion by garrypolley** [boto/botocore#2848] [Q&A] [answered]
There are some libraries that try to fill the void of missing type annotations for Python and the AWS libraries. 

Is there a reason -- other than development time and effort -- that we do not have types in this repository owned and managed by `boto`?  

If there is stomach to own/manage this I'd be happy to help draft some proposals.  I think [this line](https://github.com/boto/botocore/blob/develop/botocore/client.py#L200) is a big reason the types are not easily read/managed by code consuming boto3.  

This creates a new type dynamically which is why static type annotations don't work: 

```
type(str(class_name), tuple(bases), class_attributes)
```

Seems like the stubs could be auto generated for most combinations if someone did the work -- I'm willing to do the work if there is interest in keeping it.

**Comment by tim-finnigan** [comment, marked as answer]
Hi @garrypolley thanks for reaching out. We are tracking a feature request related to this here: https://github.com/boto/boto3/issues/1055.

I just brought this up for discussion with the team and can share the following [update](https://github.com/boto/boto3/issues/1055#issuecomment-1380848223):

>I brought this topic up for discussion with the team and the consensus was that this feature request is not currently under consideration.
>
>Integrating type hints will be considered for the next major version of boto3. In the meantime there are packages such as https://pypi.org/project/boto3-stubs/ you can use for type checking.

There is some discussion in https://github.com/boto/boto3/issues/1055 regarding the challenges of officially supporting type annotations. I don't know the full extent of the technical challenges on this front — but I do think development time and effort are a factor in why this isn't going to get prioritized in the current of boto3 (but potentially in the next major version as mentioned above).
